snakemake/deployment/spack.py

```diff
@@ -0,0 +1,232 @@
+__author__ = "Johannes Köster"
+__copyright__ = "Copyright 2021, Johannes Köster"
+__email__ = "johannes.koester@uni-due.de"
+__license__ = "MIT"
+
+import os
+import re
+import subprocess
+import tempfile
+from urllib.request import urlopen
+from urllib.parse import urlparse
+from urllib.error import URLError
+import hashlib
+import shutil
+from distutils.version import StrictVersion
+import json
+from glob import glob
+import tarfile
+import zipfile
+import uuid
+from pathlib import Path
+from enum import Enum
+import threading
+import shutil
+from .base import EnvBase
+
+from snakemake.exceptions import CreateSpackEnvironmentException, WorkflowError
+from snakemake.logging import logger
+from snakemake.common import strip_prefix, ON_WINDOWS
+from snakemake import utils
+from snakemake.deployment import singularity, containerize
+from snakemake.io import git_content
+from snakemake.sourcecache import LocalGitFile, LocalSourceFile, infer_source_file
+
+
+class SpackCleanupMode(Enum):
+    tarballs = "tarballs"
+    cache = "cache"
+
+    def __str__(self):
+        return self.value
+
+
+class Env(EnvBase):
+
+    """Spack environment from a given specification file."""
+
+    def __init__(self, env_file, workflow, env_dir=None, cleanup=None):
+
+        self.file = infer_source_file(env_file)
+        self.workflow = workflow
+
+        self._env_dir = env_dir or workflow.persistence.spack_env_path
+        self._hash = None
+        self._content_hash = None
+        self._content = None
+        self._path = None
+        self._archive_file = None
+        self._cleanup = cleanup
+
+    @property
+    def _env_archive_dir(self):
+        return self.workflow.persistence.spack_env_archive_path
+
+    @property
+    def hash(self):
+        if self._hash is None:
+            md5hash = hashlib.md5()
+            # Include the absolute path of the target env dir into the hash.
+            # By this, moving the working directory around automatically
+            # invalidates all environments. This is necessary, because binaries
+            # in spack environments can contain hardcoded absolute RPATHs.
+            env_dir = os.path.realpath(self._env_dir)
+            md5hash.update(env_dir.encode())
+            md5hash.update(self.content)
+            self._hash = md5hash.hexdigest()
+        return self._hash
```

Comment:
This looks very similar to the conda hash code. Can it be moved into the base class instead of having it in both? Please check all other methods here as well in case I missed something.

Reply:
I originally left these separate because conda has a container_img to account for (so they are different) but I can add the same variable (empty) to spack and then have some shared code to run.